### PR TITLE
Use optical photon group velocity for along step time calculation

### PR DESCRIPTION
**Pull request by hhollenb**
This PR updates the photon's simulation time be updated using the material's group velocity as determined by it's index of refraction.

The group velocity in a dispersive medium depends upon the derivative of the index of refraction with respect to frequency (equiv wrt energy). To minimally add this support without adding extra grids / calculators, I added a `DerivativeGridCalculator` which takes an `inp::Grid` and creates a new `inp::Grid` who's values are the derivatives. Because the refractive index grids are piecewise linear, the derivative is not well-defined at a grid-point. Therefore at each grid point I push it by +- epsilon so that it interpolates between the left and right derivative without exceptional behavior.

The only calculation I saw that depends on the group velocity is updating the sim time in the `AlongStepExecutor`. Because there's likely few along step actions in a volume, I guessed that it would be a better trade-off to calculate the group velocity at each step rather than adding another collection to the particle state.

### CODE REVIEW

File: test/geocel/data/optical-box.gdml [hunk 1]
```diff
@@ -4,7 +4,7 @@
   <define>
     <!-- Material properties -->
     <matrix name="RI_vacuum" coldim="2" values="1e-06 1 1e-05 1"/>
-    <matrix name="RI_water" coldim="2" values="1e-06 1.3333 1e-05 1.3333"/>
+    <matrix name="RI_water" coldim="2" values="5e-07 1 1.5e-06 1.3333 2.5e-06 1.3333 3.5e-06 2 1e-05 2"/>
```
Comment (sethrj) [inline test/geocel/data/optical-box.gdml]: Can you add tests to the optical materials for these (especially vacuum which is an "edge case")?

File: test/corecel/grid/DerivativeGridCalculator.test.cc [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+        6.8,
+        6.8,
+        191.25,
+        191.25,
+        0,
+    };
+
+    EXPECT_VEC_SOFT_EQ(expected_grid_x, deriv_grid.x);
+    EXPECT_VEC_SOFT_EQ(expected_grid_y, deriv_grid.y);
+}
```
Comment (sethrj) [inline test/corecel/grid/DerivativeGridCalculator.test.cc:61]: Can you add a test for degenerate cases such as having only two points and discontinuities from concident grid points (which is allowed)?

File: src/celeritas/optical/detail/GroupVelocityCalculator.hh [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+{
+    real_type r_index = r_index_calc_(value_as<Energy>(energy));
+    real_type r_index_deriv = r_index_deriv_calc_(value_as<Energy>(energy));
+
+    real_type group_vel
+        = constants::c_light
+          / (r_index + value_as<Energy>(energy) * r_index_deriv);
+
+    CELER_ENSURE(group_vel > 0);
+    CELER_ENSURE(group_vel < constants::c_light);
```
Comment (sethrj) [inline src/celeritas/optical/detail/GroupVelocityCalculator.hh]:
```suggestion
    CELER_ENSURE(group_vel <= constants::c_light);
```
since users will have constant `rindex=1`. Also we may want to add an assertion that rindex is positive (assuming metamaterials are not implemented in Geant4)?

File: src/corecel/grid/DerivativeGridCalculator.hh [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+ * derivative is well defined since the interpolation function is smooth.
+ *
+ *
+ * \todo Currently only linearly interpolated grids are supported since they
+ * are necessary for calculating group velocity from refractive index.
+ */
+class DerivativeGridCalculator
+{
+  public:
+    // Construct with a reference to mutable host data
```
Comment (sethrj) [inline src/corecel/grid/DerivativeGridCalculator.hh]: Copy-paste error

File: src/corecel/grid/DerivativeGridCalculator.hh [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+ *
+ * Since grids are piecewise function definitions, the derivative might not be
+ * well defined in a region around a grid-point. An epsilon-neighborhood is
+ * constructed around each grid-point where the derivative may be interpolated
+ * between the two derivative grid-points. Outside of this neighborhood the
+ * derivative is well defined since the interpolation function is smooth.
+ *
+ *
+ * \todo Currently only linearly interpolated grids are supported since they
+ * are necessary for calculating group velocity from refractive index.
```
Comment (sethrj) [inline src/corecel/grid/DerivativeGridCalculator.hh]: Assuming no spline calculation, all our grids are piecewise constant, which we can currently represent with the nonuniform grid by adding coincident points (and for which could trivially create an optimized grid calculator class that skips the interpolation). So I think if you set epsilon=0 and simplify, the result will still work.

File: src/corecel/grid/DerivativeGridCalculator.cc [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+
+        // Add right-derivative grid-point
+        result.x.push_back(grid.x[i] + epsilon_);
+        result.y.push_back(derivative(i + 1));
+    }
+    CELER_ASSERT(result.x.size() == 2 * grid.x.size());
+    CELER_ASSERT(result.y.size() == 2 * grid.y.size());
+
+    // Ensure epsilon neighborhoods don't overlap
+    CELER_ASSERT(is_monotonic_nondecreasing(make_span(result.x)));
```
Comment (sethrj) [inline src/corecel/grid/DerivativeGridCalculator.cc:65]: Because you're adding machine epsilon, rather than multiplying by (1+eps), this will likely fail when x isn't O(1).

File: test/celeritas/optical/Detector.test.cc [hunk 7]
```diff
@@ -295,13 +294,14 @@ TEST_F(DetectorTest, simple)
...
     };
-    static real_type const expected_times[] = {
-        0 + flight_time,
-        10 + flight_time,
-        1 + flight_time,
-        20 + flight_time,
-        13 + flight_time,
-        7 + flight_time,
+    // adjusted by group velocity
+    static double const expected_times[] = {
```
Comment (sethrj) [inline test/celeritas/optical/Detector.test.cc:309]: This'll need unit conversions... maybe easier/more sensible to have the test save `time * c_light / box_size`?

File: src/celeritas/optical/action/detail/AlongStepExecutor.hh [hunk 8]
```diff
@@ -41,7 +42,9 @@ CELER_FUNCTION void AlongStepExecutor::operator()(CoreTrackView& track)
     CELER_ASSERT(sim.post_step_action());
 
     // Update time
-    sim.add_time(sim.step_length() / constants::c_light);
+    auto group_vel = GroupVelocityCalculator{track.material_record()}(
+        track.particle().energy());
+    sim.add_time(sim.step_length() / group_vel);
```
Comment (sethrj) [inline src/celeritas/optical/action/detail/AlongStepExecutor.hh:48]: Can you add a little comment in the class doc about the use of the group velocity?

**Review by sethrj — COMMENTED** [review]
Looking good, thanks! We'll probably want to update the docs after #2277.

File: src/celeritas/optical/detail/GroupVelocityCalculator.hh (hunk 3 above)
Comment (hhollenb) [reply src/celeritas/optical/detail/GroupVelocityCalculator.hh]: I believe MaterialParams checks and warns against `rindex < 1` already

File: test/geocel/data/optical-box.gdml (hunk 1 above)
Comment (hhollenb) [reply test/geocel/data/optical-box.gdml]: Do we have optical materials tests somewhere? I'm not sure I see them

Comment (sethrj) [reply test/geocel/data/optical-box.gdml]: Guess not! 😅

**Comment by hhollenb** [comment]
Added a small interval in the water refractive index in `optical_box.gdml` with a constant n=1 refractive index. Checked it works with the `CELER_ENSURE(group_velocity <= constants::c_light);` and that it throws with `CELER_ENSURE(group_velocity < constants::c_light);`. Added a track in that energy interval and it does have flight time matching speed of light in a vacuum.

File: src/corecel/grid/DerivativeGridCalculator.cc [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+ * Construct the derivative grid of an imported grid.
+ *
+ * For each grid-point x in the input, the (x-epsilon, x+epsilon) is
+ * constructed. Outside the neighborhood, the derivative is well defined,
+ * whereas in the neighborhood it is interpolated between the endpoints of the
+ * neighborhood.
+ *
+ * Since only linearly interpolated grids are supported, the derivative outside
+ * the epsilon-neighborhood is just the interpolated slope between the
+ * grid-points.
```
Comment (sethrj) [inline src/corecel/grid/DerivativeGridCalculator.cc]: I think this is out of date.

File: src/corecel/grid/DerivativeGridCalculator.hh [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+ */
```
Comment (sethrj) [inline src/corecel/grid/DerivativeGridCalculator.hh]: It seems inconsistent to have the class docs in the header and function docs in the .cc file, but that matches where they're *defined*: the statement below is just a declaration.

**Review by sethrj — APPROVED** [review]
Great!

File: src/corecel/grid/DerivativeGridCalculator.hh (hunk 10 above)
Comment (hhollenb) [reply src/corecel/grid/DerivativeGridCalculator.hh]: Yeah, this was more just a remnant of the class definition.
